File: api/src/IChat.Api/Endpoints/Search/V1/Mappings/SearchMapper.cs
namespace IChat.Api.Endpoints.Search.V1.Mappings;

using IChat.Api.Endpoints.Search.V1.DTOs;
using IChat.Core.Contracts.Search;
using IChat.Core.Rag;

public static class SearchMapper
{
    public static SearchChunksRequest ToServiceRequest(this SearchRequestDto dto) =>
        new()
        {
            Query = dto.Query,
            TopK = dto.TopK,
            Mode = (SearchMode)dto.Mode,
            Rewrite = dto.Rewrite,
            History = dto.History?.Select(turn => new SearchHistoryTurn
            {
                Role = turn.Role,
                Content = turn.Content
            }).ToArray(),
            ApplyMmr = dto.ApplyMmr,
            ExpandNeighbors = dto.ExpandNeighbors,
            Rerank = dto.Rerank
        };

    public static SearchResponse ToResponse(this SearchChunksResult result) =>
        new()
        {
            OriginalQuery = result.OriginalQuery,
            RewrittenQuery = result.RewrittenQuery,
            Stages = result.Stages.ToDictionary(stage => stage.Key, stage => stage.Value.ToResponse()),
            Degraded = result.Degraded,
            ElapsedMs = result.ElapsedMs
        };

    private static SearchStageResponse ToResponse(this SearchStageView stage) =>
        new()
        {
            Name = stage.Name,
            Count = stage.Count,
            ElapsedMs = stage.ElapsedMs,
            TsQuery = stage.TsQuery,
            Top = stage.Top.Select(ToResponse).ToArray()
        };

    private static SearchHitResponse ToResponse(this SearchHit hit) =>
        new()
        {
            ChunkId = hit.ChunkId,
            DocumentId = hit.DocumentId,
            HeadingPath = hit.HeadingPath,
            Snippet = hit.Snippet,
            ChunkIndex = hit.ChunkIndex,
            Score = hit.Score
        };
}

File: ui/src/app/features/retrieval/data/search.mapper.ts
import type {
  SearchHitDto,
  SearchRequestDto,
  SearchResultDto,
  SearchStageDto,
} from '@app/core/api/retrieval.dto';
import type { SearchCriteria, SearchHit, SearchResult, SearchStage } from './search.model';

export function toSearchRequestDto(criteria: SearchCriteria): SearchRequestDto {
  return {
    query: criteria.query,
    topK: criteria.topK,
    mode: criteria.mode,
    rewrite: criteria.rewrite,
    applyMmr: criteria.applyMmr,
    expandNeighbors: criteria.expandNeighbors,
    rerank: criteria.rerank,
  };
}

function toSearchHit(dto: SearchHitDto): SearchHit {
  return {
    chunkId: dto.chunkId,
    documentId: dto.documentId,
    headingPath: dto.headingPath,
    snippet: dto.snippet,
    chunkIndex: dto.chunkIndex,
    score: dto.score,
  };
}

function toSearchStage(dto: SearchStageDto): SearchStage {
  return {
    name: dto.name,
    count: dto.count,
    elapsedMs: dto.elapsedMs,
    tsQuery: dto.tsQuery,
    top: dto.top.map(toSearchHit),
  };
}

export function toSearchResult(dto: SearchResultDto): SearchResult {
  const stages: Record<string, SearchStage> = {};

  for (const [name, stage] of Object.entries(dto.stages)) {
    stages[name] = toSearchStage(stage);
  }

  return {
    originalQuery: dto.originalQuery,
    rewrittenQuery: dto.rewrittenQuery,
    stages,
    degraded: dto.degraded,
    elapsedMs: dto.elapsedMs,
  };
}
